core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java

```diff
@@ -2354,6 +2358,72 @@ public Node visitJsonArgument(SqlBaseParser.JsonArgumentContext context)
                         .map(this::getJsonFormat));
     }
 
+    @Override
+    public Node visitJsonObject(SqlBaseParser.JsonObjectContext context)
+    {
+        List<JsonObjectMember> members = visit(context.jsonObjectMember(), JsonObjectMember.class);
+
+        boolean nullOnNull = true;
+        if (context.ABSENT() != null) {
+            nullOnNull = false;
+        }
+
+        boolean uniqueKeys = false;
+        if (context.WITH() != null) {
+            uniqueKeys = true;
+        }
+
+        Optional<DataType> returnedType = visitIfPresent(context.type(), DataType.class);
+
+        Optional<JsonFormat> jsonOutputFormat = Optional.empty();
+        if (context.FORMAT() != null) {
+            jsonOutputFormat = Optional.of(getJsonFormat(context.jsonRepresentation()));
+        }
+
+        return new JsonObject(getLocation(context), members, nullOnNull, uniqueKeys, returnedType, jsonOutputFormat);
+    }
+
+    @Override
+    public Node visitJsonObjectMember(SqlBaseParser.JsonObjectMemberContext context)
+    {
+        return new JsonObjectMember(
+                getLocation(context),
+                (Expression) visit(context.expression()),
+                (Expression) visit(context.jsonValueExpression().expression()),
+                Optional.ofNullable(context.jsonValueExpression().jsonRepresentation())
+                        .map(this::getJsonFormat));
+    }
+
+    @Override
+    public Node visitJsonArray(SqlBaseParser.JsonArrayContext context)
+    {
+        List<JsonArrayElement> elements = visit(context.jsonValueExpression(), JsonArrayElement.class);
+
+        boolean nullOnNull = false;
+        if (context.ON() != null && context.ABSENT() == null) { // should be `context.NULL() != null`, but `context.NULL()` returns an empty list when there is no `NULL`
```

Comment:
It doesn't work. `context.NULL(1) != null` would work, because with "NULL ON NULL" there are two `NULL`s. However that would be even harder to understand.

I chose
```
if (context.NULL(0) != null && context.NULL(1) != null) { // "NULL ON NULL"
```
to explicitly show that there are two nulls.